service-registry/src/main/java/org/apache/servicecomb/serviceregistry/consumer/MicroserviceVersions.java

```diff
@@ -169,6 +166,28 @@ private void setInstances(List<MicroserviceInstance> pulledInstances, String rev
     }
   }
 
+  private List<MicroserviceInstance> mergeInstances(List<MicroserviceInstance> pulledInstances,
+      List<MicroserviceInstance> inUseInstances) {
+    List<MicroserviceInstance> upInstances = pulledInstances
+        .stream()
+        .filter(instance -> {
+          return MicroserviceInstanceStatus.UP.equals(instance.getStatus());
+        })
+        .collect(Collectors.toList());
+    if (upInstances.isEmpty() && inUseInstances != null && ServiceRegistryConfig.INSTANCE.isEmptyInstanceProtectionEnabled()) {
+      MicroserviceInstancePing ping = SPIServiceUtils.getPriorityHighestService(MicroserviceInstancePing.class);
+      inUseInstances.stream()
+          .forEach(instance -> {
+            if (!upInstances.contains(instance)) {
+              if (ping.ping(instance)) {
```

Comment:
As the ping method could take some time. It could take a while if there are lots of upInstances need to be merged.

Reply:
This scenario is rarely as I explained above, mostly because of service center bugs. And this health check run a separate background thread, will not influence normal performance.